host-controller/src/main/java/org/jboss/as/host/controller/operations/AbstractDiscoveryOptionAddHandler.java

```diff
@@ -43,22 +49,42 @@
 
     public static final String OPERATION_NAME = ADD;
 
-    static void updateDiscoveryOptionsOrdering(OperationContext context, ModelNode operation,
-            String type) {
-        PathAddress operationAddress = PathAddress.pathAddress(operation.get(OP_ADDR));
-        PathAddress discoveryOptionsAddress = operationAddress.subAddress(0, operationAddress.size() - 1);
-        ModelNode discoveryOptions = Resource.Tools.readModel(context.readResourceFromRoot(discoveryOptionsAddress));
+    static void updateOptionsAttribute(final OperationContext context, final ModelNode operation, final String type)
+            throws OperationFailedException {
+        final PathAddress operationAddress = PathAddress.pathAddress(operation.get(OP_ADDR));
+        final PathAddress discoveryOptionsAddress = operationAddress.subAddress(0, operationAddress.size() - 1);
+        final ModelNode discoveryOptions = Resource.Tools.readModel(context.readResourceFromRoot(discoveryOptionsAddress));
+
+        final ModelNode node = new ModelNode();
+        if(ModelDescriptionConstants.CUSTOM_DISCOVERY.equals(type)) {
+            node.get(ModelDescriptionConstants.NAME).set(operationAddress.getLastElement().getValue());
+            for (final AttributeDefinition attribute : DiscoveryOptionResourceDefinition.DISCOVERY_ATTRIBUTES) {
+                attribute.validateAndSet(operation, node);
+            }
+        } else if(ModelDescriptionConstants.STATIC_DISCOVERY.equals(type)) {
+            node.get(ModelDescriptionConstants.NAME).set(operationAddress.getLastElement().getValue());
+            for (final AttributeDefinition attribute : StaticDiscoveryResourceDefinition.STATIC_DISCOVERY_ATTRIBUTES) {
+                attribute.validateAndSet(operation, node);
+            }
+        } else {
+            throw new OperationFailedException(HostControllerLogger.ROOT_LOGGER.invalidDiscoveryType(type));
+        }
 
         // Get the current list of discovery options and add the new discovery option to
         // this list to maintain the order
-        ModelNode discoveryOptionsOrdering = discoveryOptions.get(DISCOVERY_OPTIONS).clone();
-        if (!discoveryOptionsOrdering.isDefined()) {
-            discoveryOptionsOrdering.setEmptyList();
+        final ModelNode list = discoveryOptions.get(ModelDescriptionConstants.OPTIONS).clone();
+        if (!list.isDefined()) {
+            list.setEmptyList();
         }
-        discoveryOptionsOrdering.add(type, Util.getNameFromAddress(operation.get(OP_ADDR)));
+        list.add(type, node);
```

Comment:
This produces a list of ModelType.PROPERTY instead of the specified list of ModelType.OBJECT

```
ModelNode element = new ModelNode();
element.get(type).set(node);
list.add(element);
```

will produce the expected structure.